website/blog/2020-01-10-cobalt.md

```diff
@@ -0,0 +1,894 @@
+breakpoints in both Java and Scala source files and to check all variables at
+each.
+
+Known breakpoints limitations we are working on:
+
+- breakpoint may get ignored when set in a type containing nested type
+  declarations preceding the breakpoint (see
+  https://github.com/scalameta/metals/issues/1195)
+- breakpoint may get wrongly resolved for an ambiguous type (e.g. a.b.Foo)
+  defined inside two modules on the project (see
+  https://github.com/scalameta/metals/issues/1196)
+
+### How to add support for debugging in my editor?
+
+You will need to handle two commands in your language client extension:
+[`metals-run-session-start`](https://github.com/scalameta/metals/blob/master/metals/src/main/scala/scala/meta/internal/metals/ClientCommands.scala#L56)
+and
+[`metals-debug-session-start`](https://github.com/scalameta/metals/blob/master/metals/src/main/scala/scala/meta/internal/metals/ClientCommands.scala#L78).
+Those commands should get executed automatically by the lsp client once the user
+activates a code lens. The difference between them is that the former ignores
+all breakpoints being set while the latter respects them. The procedure of
+starting the run/debug session is as follows:
+
+1. Request the debug adapter uri from metals server using the
+   [`debug-adapter-start`](https://github.com/scalameta/metals/blob/master/metals/src/main/scala/scala/meta/internal/metals/ServerCommands.scala#L95)
+   command.
+2. Connect the debug adapter extension specific to you editor using the
+   aforementioned uri and let it drive the run/debug session. For reference,
+   take a look at the
+   [vscode implementation](https://github.com/scalameta/metals-vscode/blob/master/src/scalaDebugger.ts)
+   and how it is
+   [wired up together](https://github.com/scalameta/metals-vscode/blob/master/src/extension.ts#L356)
+
+## Worksheet support
+
+Thanks to great work by @olafurpg and some additional refinements by
+@alekseiAlefirov, worksheets are now available in all supported editors.
+
+Worksheets are way to evaluate code real time to test and prototype any new code
+in an simple way. To create a worksheet all that is needed is to create a file
+with the extension `worksheet.sc` and Metals will take care of the rest.
+
+There are two ways worksheets are implemented for different editors. One is with
+an additional extension to LSP called `Decoration extension` and it's used by
+Visual Studio Code and coc-metals.
+
+Visual Studio Code:
+
+![worksheet](https://user-images.githubusercontent.com/1408093/68093266-b0b3e680-fe8b-11e9-92dd-0165f169966f.gif)
+
+The other way worksheets are implemented for all other editors that don't
+support the `Decoration extension` is using workspace/applyEdits and comments
+with hover for larger results.
+
+Sublime:
+
+![worksheet2](https://user-images.githubusercontent.com/10850363/70265661-29fc6d00-179b-11ea-842e-935045e27f96.gif)
+
+## Better support for Vim via coc-metals
+
+TODO @ckipp01 ?
```

Comment:
In order to better support the Vim experience with Metals, [`coc-metals`](https://github.com/ckipp01/coc-metals) extension was created which is an extension for [`coc.nvim`](https://github.com/neoclide/coc.nvim), the most popular LSP client for Vim and Neovim. With this new extension comes a whole bunch of new features for Metals + Vim users.

  - automated installation
  - easier configuration options
  - easier workspace commands
  - embedded doctor
  - input box support
  - implementation of the decoration protocol (Neovim only)

![coc-metals](https://i.imgur.com/zofu4VI.png)

Reply:
Thanks, added everything here too.